### ISSUE TITLE
[Bug?] SVD will not run on integrated data

### ISSUE THREAD

**Issue by Connorr0** [timoast/signac#158]
When I have been running RunSVD I am getting an irlba error I do not regognize:

```
> integrated <- RunSVD(
+   object = integrated,
+   n = 30,
+   reduction.name = 'integratedLSI',
+ )
Running SVD
Error in irlba(A = t(x = object), nv = n, work = irlba.work) : 
  BLAS/LAPACK routine 'DLASCL' gave error code -4
```

Any ideas on where this could be coming from?

**Comment by Connorr0** [comment]
I ran it on another computer with fewer dimensions and got a new error. Raising the dimensions returns the same error.

```
> integrated <- RunSVD(
+   object = integrated,
+   n = 10,
+   reduction.name = 'integratedLSI',
+ )
Running SVD
Error in if (tol * ans$d[1] < eps) warning("convergence criterion below machine epsilon") : 
  missing value where TRUE/FALSE needed
```

**Comment by timoast** [comment]
I haven't seen this error before, can you post the full code you're running, the output of `sessionInfo()`? The error is coming from irlba, so it's most likely something to do with the integrated matrix you're using.

**Comment by Connorr0** [comment]
Here is the session info output:

R version 3.6.2 (2019-12-12)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] DropletUtils_1.6.1          EnsDb.Mmusculus.v79_2.99.0 
 [3] ensembldb_2.10.2            AnnotationFilter_1.10.0    
 [5] GenomicFeatures_1.38.2      AnnotationDbi_1.48.0       
 [7] Signac_0.2.5                biomaRt_2.42.1             
 [9] wesanderson_0.3.6           pheatmap_1.0.12            
[11] gprofiler2_0.1.8            RColorBrewer_1.1-2         
[13] viridis_0.5.1               viridisLite_0.3.0          
[15] Seurat_3.1.4                ggplot2_3.3.0              
[17] dplyr_0.8.4                 monocle3_0.2.0             
[19] SingleCellExperiment_1.8.0  SummarizedExperiment_1.16.1
[21] DelayedArray_0.12.2         BiocParallel_1.20.1        
[23] matrixStats_0.55.0          GenomicRanges_1.38.0       
[25] GenomeInfoDb_1.22.0         IRanges_2.20.2             
[27] S4Vectors_0.24.3            Biobase_2.46.0             
[29] BiocGenerics_0.32.0        

loaded via a namespace (and not attached):
  [1] reticulate_1.14          R.utils_2.9.2            tidyselect_1.0.0        
  [4] RSQLite_2.2.0            htmlwidgets_1.5.1        grid_3.6.2              
  [7] Rtsne_0.15               munsell_0.5.0            codetools_0.2-16        
 [10] mutoss_0.1-12            ica_1.0-2                future_1.16.0           
 [13] withr_2.1.2              colorspace_1.4-1         knitr_1.28              
 [16] rstudioapi_0.11          ROCR_1.0-7               gbRd_0.4-11             
 [19] listenv_0.8.0            Rdpack_0.11-1            labeling_0.3            
 [22] GenomeInfoDbData_1.2.2   mnormt_1.5-6             bit64_0.9-7             
 [25] farver_2.0.3             rhdf5_2.30.1             vctrs_0.2.2             
 [28] TH.data_1.0-10           xfun_0.12                BiocFileCache_1.10.2    
 [31] ggseqlogo_0.1            R6_2.4.1                 rsvd_1.0.3              
 [34] locfit_1.5-9.2           bitops_1.0-6             assertthat_0.2.1        
 [37] scales_1.1.0             multcomp_1.4-12          gtable_0.3.0            
 [40] npsurv_0.4-0             globals_0.12.5           sandwich_2.5-1          
 [43] rlang_0.4.4              gggenes_0.4.0            splines_3.6.2           
 [46] rtracklayer_1.46.0       lazyeval_0.2.2           reshape2_1.4.3          
 [49] tools_3.6.2              ellipsis_0.3.0           gplots_3.0.3            
 [52] ggridges_0.5.2           TFisher_0.2.0            Rcpp_1.0.3              
 [55] plyr_1.8.5               progress_1.2.2           zlibbioc_1.32.0         
 [58] purrr_0.3.3              RCurl_1.98-1.1           prettyunits_1.1.1       
 [61] openssl_1.4.1            pbapply_1.4-2            cowplot_1.0.0           
 [64] zoo_1.8-7                ggrepel_0.8.1            cluster_2.1.0           
 [67] magrittr_1.5             data.table_1.12.8        lmtest_0.9-37           
 [70] RANN_2.6.1               mvtnorm_1.0-12           ProtGenerics_1.18.0     
 [73] fitdistrplus_1.0-14      hms_0.5.3                patchwork_1.0.0         
 [76] lsei_1.2-0               XML_3.99-0.3             gridExtra_2.3           
 [79] compiler_3.6.2           tibble_2.1.3             KernSmooth_2.23-16      
 [82] crayon_1.3.4             R.oo_1.23.0              htmltools_0.4.0         
 [85] tidyr_1.0.2              RcppParallel_4.4.4       DBI_1.1.0               
 [88] dbplyr_1.4.2             MASS_7.3-51.4            rappdirs_0.3.1          
 [91] Matrix_1.2-18            R.methodsS3_1.8.0        gdata_2.18.0            
 [94] metap_1.3                igraph_1.2.4.2           pkgconfig_2.0.3         
 [97] sn_1.6-1                 GenomicAlignments_1.22.1 numDeriv_2016.8-1.1     
[100] plotly_4.9.2             dqrng_0.2.1              multtest_2.42.0         
[103] XVector_0.26.0           bibtex_0.4.2.2           stringr_1.4.0           
[106] digest_0.6.24            sctransform_0.2.1        RcppAnnoy_0.0.14        
[109] tsne_0.1-3               Biostrings_2.54.0        leiden_0.3.3            
[112] uwot_0.1.5               edgeR_3.28.1             curl_4.3                
[115] Rsamtools_2.2.2          gtools_3.8.1             lifecycle_0.2.0         
[118] nlme_3.1-142             jsonlite_1.6.1           Rhdf5lib_1.8.0          
[121] askpass_1.1              limma_3.42.2             pillar_1.4.3            
[124] lattice_0.20-38          httr_1.4.1               plotrix_3.7-7           
[127] survival_3.1-8           glue_1.3.1               png_0.1-7               
[130] bit_1.1-15.2             stringi_1.4.6            HDF5Array_1.14.3        
[133] blob_1.2.1               ggfittext_0.8.1          caTools_1.18.0          
[136] memoise_1.1.0            irlba_2.3.3              future.apply_1.4.0      
[139] ape_5.3

**Comment by Connorr0** [comment]
it started happening to my unintegrated sample as well now. Here is information on its counts matrix.

```
> sum(unintegrated@assays$peaks@counts)
[1] 41960
> sum(colSums(unintegrated@assays$peaks@counts) > 0)
[1] 328
> sum(rowSums(unintegrated@assays$peaks@counts) > 0)
[1] 12056
```

**Comment by timoast** [comment]
can you post the full code you're running?

**Comment by Connorr0** [comment]
```{r}
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(EnsDb.Mmusculus.v79)
library(dplyr)

library(DropletUtils)
sessionInfo()
```

```{r}
P0 = readRDS("P0_path")
E14 = readRDS("E14_path")

DimPlot(E14)
DimPlot(P0)
```

```{r}
#create matricies
integration_peaks_E14 <- FeatureMatrix(
  fragments = GetFragments(object = E14, assay = 'peaks'), #gets the E14 cut-sites with cell names
  features = StringToGRanges(integration_peaks, sep = c(":", "-")), # merged peaks from E14 and P0
  cells = colnames(E14)) # cells in E14

integration_peaks_P0 <- FeatureMatrix(
  fragments = GetFragments(object = P0, assay = 'peaks'),
  features = StringToGRanges(integration_peaks, sep = c(":", "-")),
  cells = colnames(P0))

length(integration_peaks)
dim(integration_peaks_E14)
dim(integration_peaks_P0)

print(paste("Total cut-sites", "E14:", sum(integration_peaks_E14)))
print(paste("Total cut-sites", "P0:", sum(integration_peaks_P0)))

E14_mean_CSpC = sum(integration_peaks_E14)/dim(integration_peaks_E14)[2]
P0_mean_CSpC = sum(integration_peaks_P0)/dim(integration_peaks_P0)[2]

print(paste("mean cut-sites per cell", "E14:", E14_mean_CSpC))
print(paste("mean cut-sites per cell", "P0:", P0_mean_CSpC))

length(integration_peaks[!(integration_peaks %in% rownames(integration_peaks_E14))])

#downsampling
# total counted cut sites e14/ total counted cut sites P0
E14_P0_frac = P0_mean_CSpC/E14_mean_CSpC

# downsamples so that there are the same number of cut-site in both samples
integration_peaks_E14_P0 <- downsampleMatrix(integration_peaks_E14, prop=E14_P0_frac)

# check number of peaks similar
P0_mean_CSpC/(sum(integration_peaks_E14_P0)/dim(integration_peaks_E14_P0)[2])


#assay assignment
E14[['mergedPeaks_E14_P0']] <- CreateAssayObject(counts = integration_peaks_E14_P0, min.cells = 0)
E14 <- RunTFIDF(object = E14, assay = 'mergedPeaks_E14_P0')

P0[['mergedPeaks_E14_P0']] <- CreateAssayObject(counts = integration_peaks_P0, min.cells = 0)
P0 <- RunTFIDF(object = P0, assay = 'mergedPeaks_E14_P0')

#visuals
hist(E14$nCount_mergedPeaks_E14_P0)
hist(P0$nCount_mergedPeaks_E14_P0)

sum(P0$nCount_mergedPeaks_E14_P0)
```

```{r}
unintegrated <- MergeWithRegions(
  object.1 = E14,
  object.2 = P0,
  assay.1 = 'mergedPeaks_E14_P0',
  assay.2 = 'mergedPeaks_E14_P0',
  sep.1 = c("-", "-"),
  sep.2 = c("-", "-")
)
dim(E14@assays$mergedPeaks_E14_P0@counts)[1]
dim(P0@assays$mergedPeaks_E14_P0@counts)[1]

unintegrated <- RunTFIDF(object = unintegrated, assay = 'peaks')
unintegrated <- FindTopFeatures(unintegrated, min.cutoff = 10)
unintegrated <- RunSVD(unintegrated, n = 30, reduction.name = 'lsi', reduction.key = 'LSI_')
unintegrated <- RunUMAP(unintegrated, reduction = 'lsi', dims = 1:30)


DimPlot(unintegrated, group.by = 'orig.ident', pt.size = 0.1) + ggplot2::ggtitle("Unintegrated E14 P0")


DimPlot(unintegrated, group.by = 'Cell_type', pt.size = 0.1) + ggplot2::ggtitle("Unintegrated E14 P0")


sum(unintegrated@assays$peaks@counts)
sum(colSums(unintegrated@assays$peaks@counts) > 0)
sum(rowSums(unintegrated@assays$peaks@counts) > 0)
```

**Comment by timoast** [comment]
Ok it looks like the reason for this is that when you run `downsampleMatrix` you can end up with some rows with 0 total counts, which causes NA values after running TF-IDF (since document frequency for those peaks is 0). NA values then causes irlba to throw that error. Simplest solution is either to not downsample the matrix (which in my opinion is a bad idea anyway), or remove rows with a row sum = 0 before running TF-IDF and SVD. I will update the TF-IDF function to issue a warning if any row or column sums are 0 and replace NA values with 0 in the output in the next release.

**Comment by Connorr0** [comment]
Awesome, thanks!

**Comment by timoast** [comment]
Fixed in Signac 1.0.0
